src/main/java/com/zaxxer/nuprocess/linux/ProcessEpoll.java

```diff
@@ -36,10 +39,19 @@
  */
 class ProcessEpoll extends BaseEventProcessor<LinuxProcess>
 {
+   private static final int EPOLL_MAX_EVENTS = 20;
    private final int epoll;
-   private final EpollEvent triggeredEvent;
+   private final EpollEvent[] triggeredEvents;
+   private final EpollEvent firstTriggeredEvent;
```

Comment:
There's a fair bit of history around the epoll code in pull request #80 that might be worth checking, as well as on commit 9df9b4ccdc055662c0406890fb1e419211ca694d (which were optimizations Brett did on top of mine).

One thing I didn't discuss on that pull request was that I profiled the difference between passing a single `EpollEvent` to `epoll_wait`  and passing an array, and I found no statistically significant difference between the two. (I wish I still had access to my write-up on the subject, but I lost that when I left Atlassian.)

I'm curious if there's something specific you're trying to accomplish with this change--and whether you have any measurements that _show_ it being accomplished. Because my own testing on this, in the past, showed there was no benefit, even under heavy process concurrency.

Ultimately, though, that's not really my primary issue with this. The real problem is you haven't accounted for `LinuxProcess.run`, only `LinuxProcess.start`. When using `start`, NuProcess's pool of long-lived `ProcessEpoll` threads handle everything. With `run`, though, _each `run` call_ uses its own short-lived `ProcessEpoll`. With this change, that means all of those `run` calls (which, again, Bitbucket Server uses _extensively_--I'd say 80-90% of its processes use `run` rather than `start`) is now allocating 20 `EpollEvent`s despite having no chance of ever needing more than 3 (in the absolute worst case; in practice I suspect it'd rarely use more than 1).

Reply:
I'll get back to you on this. In my unit test I already saw > 1 events being handled per one epoll select. I did not look into run.

In case of run(). I did just skim over the code. I do wonder. Did you ever try having a pool of selectors/epoll for the run use case? Instead of creating per process selector?